spec/unit/controllers/v3/service_instance_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+        expect(response.status).to eq 422
+        expect(response.body).to include('Unable to share to spaces')
+        expect(response.body).to include('nonexistant-space-guid')
+        expect(response.body).to include('nonexistant-space-guid2')
+        expect(service_instance.shared_spaces).to_not include(target_space)
+      end
+    end
+
+    context 'when the service instance has already been shared with the specified space' do
+      before do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+      end
+
+      it 'returns a 200 and leaves the existing share intact' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+
+        expect(response.status).to eq 200
+        expect(service_instance.shared_spaces).to include(target_space)
+      end
+    end
+
+    context 'when the request is malformed' do
+      let(:req_body) {
+        {
+          bork: 'some-name',
+        }
+      }
+
+      it 'returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+      end
+    end
+
+    describe 'permissions by role' do
+      context 'when the user is a space developer in the source space' do
+        before do
+          set_current_user_as_role(role: 'space_developer', org: source_space.organization, space: source_space, user: user)
+        end
+
+        role_to_expected_http_response = {
+          'admin'               => 200,
+          'space_developer'     => 200,
+          'admin_read_only'     => 403,
+          'global_auditor'      => 403,
+          'space_manager'       => 403,
+          'space_auditor'       => 403,
+          'org_manager'         => 403,
+          'org_auditor'         => 422,
+          'org_billing_manager' => 422,
+        }.freeze
+
+        role_to_expected_http_response.each do |role, expected_return_value|
+          context "and #{role} in the target space" do
+            it "returns #{expected_return_value}" do
+              set_current_user_as_role(role: role, org: target_space.organization, space: target_space, user: user)
+
+              post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+
+              expect(response.status).to eq(expected_return_value),
+                "Expected #{expected_return_value}, but got #{response.status}. Response: #{response.body}"
+              if expected_return_value == 200
+                expect(parsed_body['data'][0]['guid']).to eq(target_space.guid)
+              end
+            end
+          end
+        end
+      end
+
+      context 'when the user is a space developer in the target space' do
+        before do
+          set_current_user_as_role(role: 'space_developer', org: target_space.organization, space: target_space, user: user)
+        end
```

Comment:
Looks like this context and before are redundant and can be removed.

Reply:
See [comment](https://github.com/cloudfoundry/cloud_controller_ng/pull/932#discussion_r145740706) above